File: 백준/Gold/15912. 우주선 만들기/우주선 만들기.java
/**
 * Author: nowalex322, Kim HyeonJae
 */

import java.io.*;
import java.util.*;

public class Main {
    static BufferedReader br;
    static BufferedWriter bw;
    static StringTokenizer st;
    static int N, Weight[], Energy[];
    static long maxBoard[][], dp[][];
    public static void main(String[] args) throws Exception {
        new Main().solution();
    }

    public void solution() throws Exception {
        br = new BufferedReader(new InputStreamReader(System.in));
//        br = new BufferedReader(new InputStreamReader(new FileInputStream("src/main/java/BOJ_15912_우주선만들기/input.txt")));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));

        N = Integer.parseInt(br.readLine());
        Weight = new int[N];
        Energy = new int[N];
        dp = new long[N][N];
        maxBoard = new long[N][N];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < N; i++) {
            Weight[i] = Integer.parseInt(st.nextToken());
        }

        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++) {
            Energy[i] = Integer.parseInt(st.nextToken());
        }

        long maxW=0, maxE=0;
        for (int i = 0; i < N; i++) {
            maxW = Weight[i];
            maxE = Energy[i];
            for (int j = i; j < N; j++) {
                maxW = Math.max(maxW, Weight[j]);
                maxE = Math.max(maxE, Energy[j]);
                maxBoard[i][j] = maxW * maxE;
                if(i == j) dp[i][j] = maxBoard[i][j];
            }
        }

        for(int len=2; len<=N; len++) {
            for(int start=0; start<=N-len; start++) {
                int end = start+len-1;
                dp[start][end] = maxBoard[start][end]; // 한번에 구매
                for(int idx=start; idx<end; idx++) {
                    dp[start][end] = Math.min(dp[start][end], dp[start][idx] + dp[idx+1][end]); // 분할구매
                }
            }
        }

        bw.write(String.valueOf(dp[0][N-1]));
        bw.flush();
        bw.close();
        br.close();
    }
}

File: 백준/Gold/15912. 우주선 만들기/우주선 만들기.cc
/**
 * Author: nowalex322, Kim HyeonJae
 */
#include <bits/stdc++.h>
using namespace std;

// #define int long long
#define MOD 1000000007
#define INF LLONG_MAX
#define ALL(v) v.begin(), v.end()

#ifdef LOCAL
#include "algo/debug.h"
#else
#define debug(...) 42
#endif

void solve() {
    int N;
    cin >> N;
    vector<long long> W(N + 1), E(N + 1), dp(N + 1);

    for (int i = 1; i <= N; i++) {
        cin >> W[i];
    }
    for (int i = 1; i <= N; i++) {
        cin >> E[i];
    }

    for (int i = 1; i <= N; i++) {
        dp[i] = LLONG_MAX;
        long long maxW = 0, maxE = 0;
        for (int j = i; j >= 1; j--) {
            maxW = max(maxW, W[j]);
            maxE = max(maxE, E[j]);
            dp[i] = min(dp[j - 1] + maxW * maxE, dp[i]);
        }
    }

    cout << dp[N] << "\n";
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int tt = 1;  // 기본적으로 1번의 테스트 케이스를 처리
    // cin >> tt;    // 테스트 케이스 수 입력 (필요 시)

    while (tt--) {
        solve();
    }
    return 0;
}